pages/downloads/reports/conditions-report.js

```diff
@@ -0,0 +1,85 @@
+const { Router } = require('express');
+const csv = require('csv-stringify');
+const { get, omit, concat, flatten } = require('lodash');
+
+const CONDITIONS = require('@asl/projects/client/constants/conditions').default;
+const RA = require('@asl/projects/client/constants/retrospective-assessment').default;
+
+const getCondition = (condition, level = 'project', protocolTitle = '') => {
+  const defaultCondition = get(CONDITIONS[level], condition.path) || {};
+  const isCustom = condition.custom || condition.key === 'custom';
+  const isRA = condition.key === 'retrospective-assessment';
+
+  return {
+    level,
+    protocol_name: protocolTitle,
+    type: condition.type,
+    condition: isRA ? 'Retrospective Assessment' : (isCustom ? 'CUSTOM' : defaultCondition.title),
+    requires_editing: isCustom || isRA ? false : (defaultCondition.requiresEditing || false),
+    was_edited: !!condition.edited || false,
+    content: isRA ? RA.required : (condition.content || condition.edited || defaultCondition.content)
+  };
+};
+
+const getAllConditions = project => {
+  const projectConditions = (project.conditions || []).map(c => getCondition(c, 'project'));
+
+  const protocolConditions = (project.protocols || []).map(p => {
+    return (p.conditions || []).map(c => getCondition(c, 'protocol', p.title));
+  });
+
+  return concat(projectConditions, flatten(protocolConditions));
+};
+
+module.exports = settings => {
+  const router = Router({ mergeParams: true });
+
+  router.get('/', (req, res, next) => {
+
+    req.api(`/reports/ppl-conditions`)
+      .then(result => {
+        const stringifier = csv({
+          header: true,
+          columns: [
+            'establishment',
+            'licence_number',
+            'title',
+            'status',
+            'issue_date',
+            'level',
+            'protocol_name',
+            'type',
+            'condition',
+            'requires_editing',
+            'was_edited',
+            'content'
+          ]
+        });
+        res.attachment('ppl-conditions.csv');
+
+        result.json.data.forEach(project => {
+          const allConditions = getAllConditions(project);
+
+          if (allConditions.length > 0) {
+            allConditions.map((condition, i) => {
+              stringifier.write({
+                ...omit(project, 'conditions', 'protocols'),
+                ...condition
+              });
+            });
+          } else {
+            stringifier.write({
+              ...omit(project, 'conditions', 'protocols'),
+              condition: '-'
+            });
+          }
+        });
+
+        stringifier.pipe(res);
+        return stringifier.end();
+      })
+      .catch(next);
+  });
+
+  return router;
+};
```

Comment:
It feels a little bit like this is duplicating too much of the work which is already done in terms of hitting the API and CSV-ifying the response.

I'd probably extract purely the data-mapping part of it, and inject that into the existing API->CSV functionality [here](https://github.com/UKHomeOffice/asl-internal-ui/blob/master/pages/downloads/reports/index.js#L9-L10) as an optional layer.

Reply:
So that code becomes:

```js
router.get('/', (req, res, next) => {
    const report = req.params.report;
    req.api(`/reports/${report}`)
      .then(process(report))
      .then(result => {
        // stuff
      })
      .catch(next);
  });
```

And the process layer is something like:

```js
module.exports = report => records => {
  if (report === 'ppl-conditions') {
    return records.map(...); // probably extract into own file
  }
  return records;
};
```